packages/dds/merge-tree/src/test/partialLength.spec.ts

```diff
@@ -0,0 +1,174 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { MergeTree } from "../mergeTree";
+import { MergeTreeDeltaType } from "../ops";
+import { PartialSequenceLengths } from "../partialLengths";
+import { TextSegment } from "../textSegment";
+import { insertText } from "./testUtils";
+
+describe("partial lengths", () => {
+    let mergeTree: MergeTree;
+    const localClientId = 17;
+    const remoteClientId = 18;
+
+    function validatePartialLengths(
+        clientId: number,
+        seq: number,
+        expectedLen?: number,
+    ): void {
+        const partialLen = mergeTree.root.partialLengths?.getPartialLength(
+            seq,
+            clientId,
+        );
+
+        let len = 0;
+
+        mergeTree.walkAllSegments(mergeTree.root, (segment) => {
+            if (segment.isLeaf() && !segment.removedClientIds?.length) {
+                len += segment.cachedLength;
+            }
+            return true;
+        });
+
+        assert.equal(partialLen, len);
+
+        if (expectedLen !== undefined) {
+            assert.equal(partialLen, expectedLen);
+        }
+    }
+
+    beforeEach(() => {
+        PartialSequenceLengths.options.verify = true;
+        mergeTree = new MergeTree();
+        mergeTree.insertSegments(
+            0,
+            [TextSegment.make("hello world!")],
+            0,
+            localClientId,
+            0,
+            undefined);
+
+        mergeTree.startCollaboration(
+            localClientId,
+            /* minSeq: */ 0,
+            /* currentSeq: */ 0);
+    });
+
+    it("passes with no additional ops", () => {
+        validatePartialLengths(localClientId, 0);
+    });
+
+    describe("insert", () => {
+        it("local insert, local view", () => {
```

Comment:
This enumeration of cases looks reasonable. I'd ask that the test names ("describe" blocks and "it" blocks) are more sentence-like describing intended behavior.

```
describe("partial lengths", () => {
  describe("of a single inserted segment", () => {
    it("includes the segment length of a local insert, local view", () => {});

    it("doesn't include the segment length of a local insert, remote view", () => {});
  });

  describe("of a single removed segment", () => {});
});
```

etc.

I think for all of these test cases, we should probably explicitly validate length against expectations (since they're the simple cases). Then for more complicated ones we can just trust the computation.

Then, we probably want another set of describe blocks for interesting partial length aggregation cases. For example, if you have an entire merge block full of segments in each of the 8 states you've identified (at 8 different seq numbers), the partial length object should still work correctly at all of those 8 seq values. I see this as useful to test on top of your current tests because the logic that happens in PartialLengths is meaningfully different (see PartialLengths.combine)

There's a useful pairwise tool that Tony wrote which might be useful here for even more complex cases.